### ISSUE TITLE
How can I use a LocalSet within a Send future to ensure LocalKey::with will be the same

### ISSUE THREAD

**Issue by Bajix** [tokio-rs/tokio#4141]
Here's an annotated example pulled from a very rough proof of concept I was [contemplating](https://github.com/Bajix/deque-loader-rs/commit/06d6c53860a3fc8670a8bc9c805344e0167af206). The issue that i'm struggling with is that it would so seem load_by becomes !Send even though it doesn't matter which thread executes this, only that the spawn_locals occur within the same thread. Is there a different design pattern that can accomplish this with tokio? I could rewrite my algorithm to have Send stealers using crossbeam epoch if this is nonviable, or I could rewrite this to schedule all loads and cell swaps by using thread local async_executor instances. 

```rust
  fn push(loader: &'static LocalKey<DataLoader<T>>, req: Request<T::Key, T::Value, T::Error>) {
  // This loader *needs* to be the same loader as the later as called within spawn_local
  if loader.with(|loader| loader.queue.push(req)).eq(&0) {
      // By making this schedule within the same thread, this ensures that the batch encompasses all other loads that are schedule ahead of it... I think
      tokio::task::spawn_local(async move {
        // This swaps the buffer cell and converts it back into a Vec
        let requests = loader.with(|loader| loader.queue.take_batch());

        let task = Task::new(requests);

        // It doesn't matter what thread runs this; the only thing that is important is that we take from the same cell
        T::handle_task(task).await;
      });
    }
  }

  pub async fn load_by(
    loader: &'static LocalKey<DataLoader<T>>,
    key: T::Key,
  ) -> Result<Option<Arc<T::Value>>, T::Error> {
    let (req, rx) = Request::new_oneshot(key);

    let local = tokio::task::LocalSet::new();

    local.spawn_local(async move {
      // All that matters is that the first push triggers a spawn_local that can access the same Cell using LocalKey::with
      DataLoader::push(loader, req);
    });

    local.await;

    let data = rx.recv().await?;

    Ok(data)
  }
```

```
   = help: within `impl warp::Future`, the trait `std::marker::Send` is not implemented for `*const ()`
   = note: required because it appears within the type `PhantomData<*const ()>`
   = note: required because it appears within the type `LocalSet`
```

**Comment by Darksonn** [comment]
You can't use a `LocalSet` from within a task spawned with `tokio::spawn`. The documentation for `LocalSet` explains this. However I don't understand why you're trying to do that — a `LocalKey` should work even if the task is migrated between threads.

**Comment by Bajix** [comment]
> You can't use a `LocalSet` from within a task spawned with `tokio::spawn`. The documentation for `LocalSet` explains this.

As written this doesn't use LocalSet from within `tokio::task::spawn`.

> However I don't understand why you're trying to do that — a LocalKey should work even if the task is migrated between threads.

I'm playing around with different concepts for batching load operations. Previously I was doing this where whenever the first load `Request` was pushed to a thread local Crossbeam Deque that this would then trigger `tokio::task::spawn` and I would work-steal drain this into a batch that could then be loaded together on a single database query against the key set, and this works fairly well, but I think this can be further optimized and so that's what I'm endeavoring to do. After considering this for a while, I got the idea that as an alternate design, I could use atomics to write volatile across a succession of slots within a resizable buffer stored in a Cell, and then this could be later swapped and the buffer converted back into a Vec; basically crossbeam deque without the pop behavior and more around the swap. The problem here that I now need to solve for is how to take that batch at the right time, which was why I was looking into LocalSet as a solution thinking that `spawn_local` would be scheduled after my existing loads and therefor be the correct time to take the batch. I think probably an alternate design I could do is to make it so that my data loaders own async_executor LocalExector's and that I use this solely for awaiting my load receivers and scheduling the tasks.

**Comment by Darksonn** [comment]
> As written this doesn't use LocalSet from within `tokio::task::spawn`.

Well, you would not get that error if you didn't use it from `tokio::spawn`. The incomplete error message you posted seems to mention warp, and warp will generally `tokio::spawn` for you. (Please post full error messages in the future.)

I still don't really understand what you are trying to do, but my advice is don't try to use a `LocalSet` or a `LocalExecutor` in this way. (Unrelated, but there's generally no point in using `LocalExecutor` if you can use `LocalSet`.) The `LocalKey` primitive works only within a single async task, but `spawn_local` will spawn new tasks and hence it will not be transferred across the task boundary. Find another way to share the value, e.g. maybe use an `Arc`.

**Comment by Bajix** [comment]
> Well, you would not get that error if you didn't use it from `tokio::spawn`. The incomplete error message you posted seems to mention warp, and warp will generally `tokio::spawn` for you. (Please post full error messages in the future.)

Ah fair point I hadn't considered that

> I still don't really understand what you are trying to do, but my advice is don't try to use a LocalSet or a LocalExecutor in this way. (Unrelated, but there's generally no point in using LocalExecutor if you can use LocalSet.) The LocalKey primitive works only within a single async task, but spawn_local will spawn new tasks and hence it will not be transferred across the task boundary. Find another way to share the value, e.g. maybe use an Arc.

The challenge actually isn't how to share the value; it's how to time things so that optimal batching will occur. Say for instance I load 50 posts, and then load 50 authors by ID for each of these posts; the first author load schedules the task, and the other 49 simply push so that when the task runs, we can now take the entire batch of 50. If I use a task stealer approach similar to crossbeam deque to make thread safe stealers and spawn, then in all likelihood this spawn will be ran on the same thread and the batch will be 50, and definitely all will be processed, but it won't be with absolute certainty that all 50 are collected together. If however I can do this in a way in which I can guarantee that the first load triggers a local future, then in this case the other 49 loads will have an opportunity to push to my thread local buffer before this future then executes and I can take 50 in a single batch at the soonest possible time and with it being deterministic that all 50 will be batched together. If I just do a normal spawn, then it's not deterministic that this batches all 50 because another thread could steal the spawn and run it earlier than expected, though in practice this doesn't really happen and worst-case could result in a set of resolvers from the same GraphQL request having loads batched independently under heavy load. I think that since LocalSet won't work within Warp no matter what, then I can use `LocalExector`, and the only thing this would give me is the ability to deterministically collect all observed pushes within a timeout-free, unbounded logical batch. I think with having an &'static `std::thead::LocalKey`, then it's my understanding that I can pass around this reference freely and it'll only actually resolve to the current thread when using `with`, which would mean that if I use this with `LocalExecutor` I wouldn't need to Rc my Cell. In any case, I think that the design goal around making this thread-local is more around take-timing than it is about wanting to use Cell for the sake of using Cell. I can absolutely use crossbeam::epoch::Atomic to the same effect, but that still wouldn't give me the *perfect* timing for batch taking, which is the ultimate goal

**Comment by Darksonn** [comment]
It sounds like you want to ensure that several things are executed on the same thread? You can do this with [`FuturesUnordered`][1] or [`StreamExt::buffer_unordered`][2] or [`StreamExt::buffered`][3]. All of these combinators guarantee that everything you put inside them run in the same task, and hence also on the same thread.

Regarding the other things, if you are in a situation where you can't use a `LocalSet`, then a `LocalExector` wont work either. I think you should entirely forget all about `LocalSet`, `LocalKey` and `LocalExecutor` here. I strongly doubt that they are what you are looking for.

[1]: https://docs.rs/futures/0.3/futures/stream/struct.FuturesUnordered.html
[2]: https://docs.rs/futures/0.3/futures/stream/trait.StreamExt.html#method.buffer_unordered
[3]: https://docs.rs/futures/0.3/futures/stream/trait.StreamExt.html#method.buffered

**Comment by Bajix** [comment]
Those wouldn't be well suited here as the goal is to collect a bunch of key/receiver pairs and then to query against the key set and to load/resolve together with each of these loads will being in a different GraphQL field resolver. The problem this is solving for is accomplish context-free optimal batching purely by clever timing by relying on field resolvers pushing all loads in a logic batch before our eventual callback takes this batch. 

What makes you say that LocalExecutor wouldn't work here? If executors are owned by thread local data loaders then this could facilitate timing-based batching potentially. I'm not seeing why this isn't viable, but as a next-best option I could pursue a crossbeam::epoch::Atomic approach for facilitating thread-safe buffer swaps

**Comment by Darksonn** [comment]
`LocalExecutor` is not `Send`, so if you are in a situation where `LocalSet` doesn't work, then neither will `LocalExecutor`. And in reverse, if you are in a situation where you _can_ use `LocalExecutor`, then you might as well use a `LocalSet` there, avoiding introducing an extra crate as Tokio already provides an executor.

This question sounds like a multi-layered [XY problem][1]. Your original question was about how to combine `LocalSet` with `LocalKey`, for which the answer that you can't. Now you are talking about some sort of timing question, and I don't really understand what you mean, but it also does not sound like the actual underlying problem you are trying to solve. What is the actual problem you are trying to solve by doing this fancy timing stuff?

[1]: https://xyproblem.info/

**Comment by Bajix** [comment]
> LocalExecutor is not Send, so if you are in a situation where LocalSet doesn't work, then neither will LocalExecutor. And in reverse, if you are in a situation where you can use LocalExecutor, then you might as well use a LocalSet there, avoiding introducing an extra crate as Tokio already provides an executor.

It wouldn't matter that LocalExecutor isn't `Send`; in my current design, `crossbeam::deque::Worker` also isn't `Send` and it's only the workers stealer that is `Send`. It sounds like from everything you've said, it's not viable for me to use `LocalSet` because of the constraint that it can't be used within `tokio::task::spawn` and therefor cannot be used with warp ever in the way I was conceiving. If I used a thread local `LocalExecutor` that I used only for load receivers and batch takes then this would give me the ability to spawn onto that executor for the first load while using run_until on all oneshot receivers and in theory the spawn will execute only after all the oneshot senders have been pushed into my buffer so that then I can swap the buffer, convert it to a Vec, batch load, and resolve all the receivers. I think by design the only things that would need to be on the `LocalExecutor` are the oneshot receiver awaits and the detached spawn that collects the corresponding senders. So it's plausible there's a solution to be had, and it would so seem that there's no way this can be done with Tokio that works with warp because of this limitation on `LocalSet`.

> This question sounds like a multi-layered XY problem. Your original question was about how to combine LocalSet with LocalKey, for which the answer that you can't. Now you are talking about some sort of timing question, and I don't really understand what you mean, but it also does not sound like the actual underlying problem you are trying to solve. What is the actual problem you are trying to solve by doing this fancy timing stuff?

If I have a GraphQL request that calls a bunch of field resolvers and each of those field resolvers loads using a specific data loader by an ID, I need a mechanism to logically batch these together so that I can do a single load for all keys and then to resolve each oneshot receiver. I'm trying to create a solution that is better than using a thread-local `crossbeam::deque::Worker` and work stealing within a tokio::task::spawn spawned upon the first push; the two ways I think this can be improved upon would be to steal all at once via a buffer swap, and to make the spawn always local. I was exploring the latter concept first because should I solve for doing the buffer swap at the right time and only using Cell, then presumably this will be simpler and more performant than using crossbeam::epoch::Atomic as my Cell alternative.

**Comment by Darksonn** [comment]
The point is that the solution you describe with `LocalExecutor` would also work with a `LocalSet` instead of the `LocalExecutor`.

---

So it sounds like you have GraphQL requests being emitted from separate tasks, but you want them to be collected together and batched automatically?

**Comment by Bajix** [comment]
> The point is that the solution you describe with `LocalExecutor` would also work with a `LocalSet` instead of the `LocalExecutor`.
> 
> So it sounds like you have GraphQL requests being emitted from separate tasks, but you want them to be collected together and batched automatically?

Yea; it doesn't matter if the loads come from the same GraphQL request or not, this batches everything together in a thread local manner. This needs to not batch prematurely for a single request that has multiple field resolvers that can be batched together, yet to still batch together in the fastest manner. It's conceptually similar to doing a NodeJS process.nextTick data loader that collects everything within a single event loop.

**Comment by Darksonn** [comment]
Here's a rather different suggestion: Create an actor that is responsible for actually sending the GraphQL requests, using the pattern outlined in [this blog post][1]. This actor will put requests into a queue when they are received, and then put a timeout on the recv call, starting a new GraphQL request whenever the timeout elapses or when the vector becomes large.

With this method, you can even batch stuff happening on different threads, which is most likely worth it since thread-synchronization is extremely cheap compared to network calls.

[1]: https://ryhl.io/blog/actors-with-tokio/

**Comment by Bajix** [comment]
The problem with this approach, which I've already previously tried out
fairly thoroughly, is that you incur a compounding timeout overhead and the
batching isn't as performant even if all you do is async receive one
message and try_recv the remainder. Doing this with a single global batch
and found that this doesn't work well under heavy load, and so the current
implementation groups into worker groups such that batching will occur via
work stealing the entire group at a time. My earlier tests with channels
weren't as performant as the current implementation using
`crossbeam::deque::Worker` either. So right I think that taking the entire
queue at once via swapping the buffer is the only way I can improve upon
the performance when compared to `crossbeam::deque::Worker`, which itself
is more performant than using Flume/Tokio channels. I think that a
thread-local design scales better under heavy load, and if I want to do
stealing across multiple threads I can probably do that by stealing batches
instead of individual loads by using the swap-buffer based approach to
batch within logical time frames and pushing this Vec onto a Deque to be
stolen from any thread.

**Comment by Darksonn** [comment]
You may want to look into doing something similar to described [here](https://medium.com/@fujita.tomonori/scalable-server-design-in-rust-with-tokio-4c81a5f350a3).

Anyway, I am closing this issue as it is not a Tokio bug or feature request.

**Comment by Bajix** [comment]
I'll take a look. Do you think it'll ever be the case that LocalSet will be
usable within tokio::task::spawn or that tokio::task::spawn_local will be
able to be used outside of LocalSet?

**Comment by Darksonn** [comment]
There's been some discussion recently about introducing a new runtime flavor without work stealing, which would make `spawn_local` possible on that runtime without a `LocalSet`.

Using `LocalSet` within an ordinary `tokio::spawn` will never be possible.

**Comment by Bajix** [comment]
I could see that offering quite a few performance optimizations as then the runtime itself could leverage `!Send` primitives as building blocks. I recall in some discussion it mentioned that the Actix runtime uses thread-per-core Tokio single-threaded Runtimes and is able to perform better because of this but I haven't really given that too much thought. 

Perhaps there's a middle ground here: a runtime without work stealing that use something like tokio::task::spawn_detached as a new behavior that can designate a task as something that can be lazily ran by any thread as opportunistically permitted.

**Comment by Bajix** [comment]
Just as a follow up in case you're curious: I figured out how to make this [data structure](https://github.com/Bajix/swap-queue-rs) capable of having thread-safe stealer handles and so the !Send aspect of this queue design is no longer problematic. I also tried `LocalExecutor`, but the `!Send` aspect of that makes it invariably impossible to use outside of main or within a thread, and I just didn't realize at first that this couldn't be solved through thread local use because I was already successfully using thread local `!Send` deques, but for that there's no async boundary so yea, you were so right. So I guess now I can do my concept with `tokio::task::spawn` and it'll sometimes be less-than-ideal when this task is stolen, but if Tokio ever does make !Send futures more usable I'll revisit this just to see if this makes for better batch-take timing.
